himbaechel/uarch/gowin/gowin.cc

```diff
@@ -725,6 +736,73 @@ void GowinImpl::assign_cell_info()
     }
 }
 
+// If there is an unused LUT next to the DFF, use its inputs for the D input
+void GowinImpl::create_passthrough_luts(void)
+{
+    std::vector<std::unique_ptr<CellInfo>> new_cells;
+    // evenly use all 4 LUT inputs
+    const std::vector<std::pair<IdString, int>> lut_tmpl = {
+            {id_I0, 0xaaaa}, {id_I1, 0xcccc}, {id_I2, 0xf0f0}, {id_I3, 0xff00}};
+    int cur_lut_tmpl = 3;
+
+    for (auto &cell : ctx->cells) {
+        CellInfo *ci = cell.second.get();
+        if (is_dff(ci)) {
+            Loc loc = ctx->getBelLocation(ci->bel);
+            BelId lut_bel = ctx->getBelByLocation(Loc(loc.x, loc.y, loc.z - 1));
+            CellInfo *lut = ctx->getBoundBelCell(lut_bel);
+            CellInfo *alu = ctx->getBoundBelCell(ctx->getBelByLocation(Loc(loc.x, loc.y, loc.z / 2 + BelZ::ALU0_Z)));
+            const CellInfo *ramw = ctx->getBoundBelCell(ctx->getBelByLocation(Loc(loc.x, loc.y, BelZ::RAMW_Z)));
+
+            if (!(lut || alu || ramw)) {
+                if (ctx->debug) {
+                    log_info("Found an unused LUT:%s, ", ctx->nameOfBel(lut_bel));
+                }
+                // make LUT
+                auto lut_cell = gwu.create_cell(gwu.create_aux_name(ci->name, 0, "_passthrough_lut$"), id_LUT4);
+                CellInfo *lut = lut_cell.get();
+                NetInfo *d_net = ci->getPort(id_D);
+                NPNR_ASSERT(d_net != nullptr);
+
+                if (d_net->name == ctx->id("$PACKER_GND") || d_net->name == ctx->id("$PACKER_VCC")) {
+                    if (ctx->debug) {
+                        log("make a constant %s.\n", d_net->name == ctx->id("$PACKER_VCC") ? "VCC" : "GND");
+                    }
+                    ci->disconnectPort(id_D);
+                    if (d_net->name == ctx->id("$PACKER_GND")) {
+                        lut->setParam(id_INIT, 0x0000);
+                    } else {
+                        lut->setParam(id_INIT, 0xffff);
+                    }
+                } else {
+                    if (ctx->debug) {
+                        log("make a pass-through.\n");
+                    }
+                    IdString lut_input = lut_tmpl.at(cur_lut_tmpl).first;
```

Comment:
is there a reason for changing which LUT input is used each time? this seems like something that long term might be better dealt with using LUT permutation

Reply:
There is no particular reason. I can use the I3 input to all LUTS if you think it is so convenient for the one who will implement Permutation in the future.

Reply:
It would make the code a bit simpler, if there are no other downsides I would just stick to I3 particularly if permutation is planned for the future

Reply:
done